## Problem #124: Ordered Radicals
[Link to Problem](https://projecteuler.net/problem=124)

### Problem Description

The radical of $n$, $\operatorname{rad}(n)$, is the product of the distinct prime factors of $n$. For example, $504 = 2^3 \times 3^2 \times 7$, so $\operatorname{rad}(504) = 2 \times 3 \times 7 = 42$.

If we calculate $\operatorname{rad}(n)$ for $1 \le n \le 10$, then sort them on $\operatorname{rad}(n)$, and sorting on $n$ if the radical values are equal, we get:

<table class="center">
<tr>
   <th colspan="2">Unsorted</th>
   <th colspan="3">Sorted</th>
</tr>
<tr>
   <th class="w50"><i>n</i></th>
   <th class="w50">rad(<i>n</i>)</th>
   <th class="w50"><i>n</i></th>
   <th class="w50">rad(<i>n</i>)</th>
   <th class="w50">k</th>
</tr>
<tr>
   <td>1</td><td>1</td>
   <td>1</td><td>1</td><td>1</td>
</tr>
<tr>
   <td>2</td><td>2</td>
   <td>2</td><td>2</td><td>2</td>
</tr>
<tr>
   <td>3</td><td>3</td>
   <td>4</td><td>2</td><td>3</td>
</tr>
<tr>
   <td>4</td><td>2</td>
   <td>8</td><td>2</td><td>4</td>
</tr>
<tr>
   <td>5</td><td>5</td>
   <td>3</td><td>3</td><td>5</td>
</tr>
<tr>
   <td>6</td><td>6</td>
   <td>9</td><td>3</td><td>6</td>
</tr>
<tr>
   <td>7</td><td>7</td>
   <td>5</td><td>5</td><td>7</td>
</tr>
<tr>
   <td>8</td><td>2</td>
   <td>6</td><td>6</td><td>8</td>
</tr>
<tr>
   <td>9</td><td>3</td>
   <td>7</td><td>7</td><td>9</td>
</tr>
<tr>
   <td>10</td><td>10</td>
   <td>10</td><td>10</td><td>10</td>
</tr>
</table>

Let $E(k)$ be the $k$-th element in the sorted $n$ column; for example, $E(4) = 8$ and $E(6) = 9$.

If $\operatorname{rad}(n)$ is sorted for $1 \le n \le 100000$, find $E(10000)$.

### Approach

Compute the smallest prime factor for each number $\leq 10^5$. This will then help us compute $rad(n)$ for any $n$ up to $10^5$. Then we just sort the numbers.

In [4]:
L = 10 ** 5 + 1

spf = [i for i in range(L)]

for i in range(2, int(L ** 0.5) + 1):
    if spf[i] == i:
        for j in range(i * i, L, i):
            spf[j] = min(spf[j], i)

rad = [1] * L

for i in range(2, L):
    if spf[i] == spf[i // spf[i]]:
        rad[i] = rad[i // spf[i]]
    else:
        rad[i] = rad[i // spf[i]] * spf[i]

order = sorted(range(L), key=lambda x: (rad[x], x))

print(order[10000])

21417


###### Result: **21417** | Execution time: ~0s

### Complexity analysis

Time complexity: $O(N\log{N})$ from the sieve algorithm for finding spf and from the sorting

##### Tags: #primes, #sieve, #sorting